lib/ddtrace/contrib/extensions.rb

```diff
@@ -14,20 +15,33 @@ module Contrib
     # process.
     # Most of this file should probably live inside the tracer core.
     module Extensions
-      def self.extended(base)
-        base.extend(Helpers)
-        base.extend(Configuration)
+      def self.extend!
+        Datadog::Tracing.singleton_class.prepend Helpers
+        Datadog::Tracing.singleton_class.prepend Configuration
+        Datadog::Configuration::Settings.include Configuration::Settings
+        Datadog::Configuration::ValidationProxy::Tracing.include Configuration::ValidationProxy
       end
```

Comment:
I think we're going in the wrong direction here. Needing to monkey patch our core to support this file is I think a very strong hint that this part of the code (the registry) should belong to the tracer core.

Note that this doesn't mean the full contrib would be part of the core -- just the bits that need to exist to actually support contrib.

This would mean that every module here would be merged back to the core class, and thus also would simply looking at and testing those classes: we'd get WYSIWYG when looking at the sources, rather than needing to "remember" that there's additional hidden behavior that is monkey patched from the outside.

Reply:
> This manifests as problematic in the future such that if core and contrib were their own packages, I would have to update two packages in synchronization instead of just one. contrib should act as a public usage of the core, not a privileged insider. The only behavior that contrib really adds to core is c.instrument behavior. That's pretty much it. The rest is non-invasive public API usage.

To be honest, this argument could be used either way. Right now this is my exact concern: we're monkey patching the core, and so if we the core changes, so would the contrib package need to.

For instance, consider pre-1.x ddtrace: if we had already separated the contrib to a separate package, it would be broken by the 1.x changes and moves to the configuration! This is I believe a strong indicator that this should live in the core.

> And behavior isn't (and shouldn't be) hidden either; it should be tested from the Contrib side of tests, since its the one doing the injection.

Since we're monkey patching things, this is actually not the case -- everything gets affected by the additions, not just the contrib side. If there's a bug in the contrib additions, for instance forgetting to delegate to super, everything gets affected, not just the contrib.

I believe this qualifies as [breaking the "O" in SOLID](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle): changes done by the contrib module can break/disturb existing users of those classes, because we're injecting behavior (monkey patching) them directly.

Reply:
> For instance, consider pre-1.x ddtrace: if we had already separated the contrib to a separate package, it would be broken by the 1.x changes and moves to the configuration!

Yes, that's what I consider working as intended. This is the point of semantic versioning; we are allowed to break other packages as long as we bump a major version. If anything, it would have made this whole 1.0 release much quicker and smaller, because we could have released the core changes first, then updated the contrib on its own afterwards, rather than all at once in a mega-monolith release. The example only reinforces my belief that it should be separated.

> Since we're monkey patching things, this is actually not the case -- everything gets affected by the additions, not just the contrib side.

I would say the same argument would apply to `ddtrace` as a whole, given we have to monkey patch other OSS libraries. We affect their code and its behavior, yet there's no need to embed our tracing into their repos/packages directly.

If we're doing more re-org later, we can discuss this again, but for the time being I have no intention on re-architecting the `Contrib` namespace for the 1.0 release. It's yet another expansion of scope and delaying the completion of work.

Reply:
Indeed, sorry if I was unclear: I'm not advocating for this hypothetical bigger change to be made in this PR, just that I think we should reconsider the current direction for the reasons I presented.

I brought this up in this PR because we're actively changing and touching this class, so I think it's relevant to the ongoing discussion, to establish a common vision of where we're going.